server/genquery2/src/genquery2_sql.cpp

```diff
@@ -0,0 +1,1172 @@
+
+#include <boost/graph/graph_traits.hpp>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/iterator/function_input_iterator.hpp>
+
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <array>
+#include <map>
+#include <stdexcept>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+namespace
+{
+    namespace gq = irods::experimental::api::genquery2;
+
+    // clang-format off
+    using graph_type         = boost::adjacency_list<boost::vecS,
+                                                     boost::vecS,
+                                                     boost::undirectedS,
+                                                     irods::experimental::genquery::vertex_property,
+                                                     irods::experimental::genquery::edge_property>;
+    using vertex_type        = boost::graph_traits<graph_type>::vertex_descriptor;
+    using vertices_size_type = boost::graph_traits<graph_type>::vertices_size_type;
+    using edge_type          = std::pair<vertex_type, vertex_type>;
+
+    using log_gq             = irods::experimental::log::genquery2;
+    // clang-format on
+
+    struct gq_state
+    {
+        // Holds pointers to column objects which contain SQL CAST syntax.
+        // These pointers allow the parser to forward the SQL CAST text to the final output.
+        std::vector<const gq::column*> ast_column_ptrs;
+
+        std::vector<std::string_view> sql_tables;
+        std::map<std::string, std::string> table_aliases;
+        std::vector<std::string> values;
+
+        int table_alias_id = 0;
+
+        bool in_select_clause = false;
+
+        bool add_joins_for_meta_data = false;
+        bool add_joins_for_meta_coll = false;
+        bool add_joins_for_meta_resc = false;
+        bool add_joins_for_meta_user = false;
+
+        bool add_sql_for_data_resc_hier = false;
+    }; // struct gq_state
+
+    // clang-format off
+    constexpr auto table_names = std::to_array({
+        "R_COLL_MAIN",                  // 0
+        "R_DATA_MAIN",                  // 1
+        "R_META_MAIN",                  // 2
+        "R_OBJT_ACCESS",                // 3
+        "R_OBJT_METAMAP",               // 4
+        "R_RESC_MAIN",                  // 5
+        "R_RULE_EXEC",                  // 6
+        "R_SPECIFIC_QUERY",             // 7
+        "R_TICKET_ALLOWED_HOSTS",       // 8
+        "R_TICKET_ALLOWED_USERS",       // 9
+        "R_TICKET_ALLOWED_GROUPS",      // 10
+        "R_TICKET_MAIN",                // 11
+        "R_TOKN_MAIN",                  // 12
+        "R_USER_AUTH",                  // 13
+        "R_USER_GROUP",                 // 14
+        "R_USER_MAIN",                  // 15
+        "R_USER_PASSWORD",              // 16
+        "R_USER_SESSION_KEY",           // 17
+        "R_ZONE_MAIN",                  // 18
+        "R_QUOTA_MAIN",                 // 19
+    }); // table_names
+    // clang-format on
```

Comment:
Could be useful in `server/core/include/irods/catalog_utilities.hpp`.

Reply:
Maybe. It's not clear to me what this would be used for though. I think we let it ride for now unless you or someone sees use-cases for moving it.

Reply:
Don't really have any use cases in mind. We just have a bunch of catalog-related information in there and having all the table names seems like it could be useful. Feel free to just resolve this.

Reply:
Ok. Resolving.

Reply:
Once GenQuery2 becomes stable, we can revisit this. I think it's too early to make this part of the other header file. I may have made some assumptions about that table in regard to the parser implementation.